haystack/preview/components/websearch/serper_dev.py

```diff
@@ -0,0 +1,138 @@
+import json
+import logging
+from typing import Dict, List, Optional, Any
+
+import requests
+
+from haystack.preview import Document, component, default_from_dict, default_to_dict
+from haystack.preview.components.websearch.utils import score_results
+
+logger = logging.getLogger(__name__)
+
+
+@component
+class SerperDev:
+    """
+    Search engine using SerperDev API. See the [Serper Dev website](https://serper.dev/) for more details.
+    """
+
+    def __init__(
+        self,
+        api_key: str,
+        top_k: Optional[int] = 10,
+        allowed_domains: Optional[List[str]] = None,
+        search_params: Optional[Dict[str, Any]] = None,
+    ):
+        """
+        :param api_key: API key for the SerperDev API.
+        :param top_k: Number of documents to return.
+        :param allowed_domains: List of domains to limit the search to.
+        :param search_params: Additional parameters passed to the SerperDev API.
+        For example, you can set 'num' to 20 to increase the number of search results.
+        """
+        if api_key is None:
+            raise ValueError("API key for SerperDev API must be set.")
```

Comment:
In `ChatGPTGenerator` I've made possible to set the API key either in `__init__` or in `run`. It's not necessary to do the same here, but it might be convenient: for example, it makes easy for the users to not have the API key serialized with the component if they don't want to. @silvanocerza do you have any opinion on this?

Reply:
I'd rather we add serialization field exclusion for such cases? We are kinda of avoiding that and stuffing api_key in run signature as a side-effect, right? @silvanocerza

Reply:
I think setting the key in `run` is not necessary, why would I want to change it while running my component?

Secrets serialization instead is something that I'm thinking about. Certainly a component's secrets must not be saved to file for a plethora of reasons, since each component knows the information that it can save this could be handled by it.

I think the better solution would be something similar to this. When saving to file we set the secret value to an env var that will be used when loading back from file. This way we avoid leaking secrets but also versatile enough.

```
    def to_dict(self):
        return {"my_secret": "COMPONENT_SECRET_VALUE", "other_field": self.other_field}
    
    @classmethod
    def from_dict(cls, data):
        data["my_secret"] = os.environ[data["my_secret"]]
        return cls(data)
```

Reply:
> I think setting the key in run is not necessary, why would I want to change it while running my component?

I was thinking that more than one user might want to use the same pipeline, like in a simple rest api demo or something like that. Definitely not a priority right now but I see some future use.